### PR TITLE
Zext sext undef in sdag

### PR DESCRIPTION
**Pull request by fengfeng09**
extend an undef value in sdag should keep the undef info.

### CODE REVIEW

**Comment by RKSimon** [comment]
What is the purpose behind this?

This approach is incorrect for how we handle extensions of undefs: https://alive2.llvm.org/ce/z/A4X7D5

**Comment by fengfeng09** [comment]
> What is the purpose behind this?
> 
> This approach is incorrect for how we handle extensions of undefs: https://alive2.llvm.org/ce/z/A4X7D5

In fact, there is no practical significance for now. But it is counterintuitive to evaluate extend undef to a zero. Especially, when we use undef value as a element in BUILD_VECTOR, it may produce some unexpected zeros.

File: llvm/test/CodeGen/X86/zext-sext.ll [hunk 1]
```diff
@@ -77,5 +77,15 @@ entry:
   %alphaXbetaY = add i64 %alphaX, %tmp115
   %transformed = add i64 %alphaXbetaY, 9040145182981852475
   store i64 %transformed, ptr %d, align 8
+  %tmp200 = zext i16 undef to i32
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/zext-sext.ll:80]: This test doesn't show anything

Comment (fengfeng09) [reply llvm/test/CodeGen/X86/zext-sext.ll:80]: two 16 bit undef composed a zero？

**Comment by topperc** [comment]
This optimization is not correct for 'undef'. I think it would be correct for 'poison', but SelectionDAG doesn't have 'poison' today.

See also https://blog.regehr.org/archives/1837

**Comment by fengfeng09** [comment]
> This optimization is not correct for 'undef'. I think it would be correct for 'poison', but SelectionDAG doesn't have 'poison' today.
> 
> See also https://blog.regehr.org/archives/1837

I see. Thanks for your explanation, so zext undef to a zero is by designed, is this right?

**Comment by topperc** [comment]
> > This optimization is not correct for 'undef'. I think it would be correct for 'poison', but SelectionDAG doesn't have 'poison' today.
> > See also https://blog.regehr.org/archives/1837
> 
> I see. Thanks for your explanation, so zext undef to a zero is by designed, is this right?

For (zext undef), the zero extended bits of the result must be 0. The bits that started out as undef can be any value. We choose 0 for simplicity.

For (sext undef), the sign extended bits must have the same value as the original sign bit. Because of undef, the original sign bit can be either 0 or 1. Folding to undef wouldn't guarantee the extended bits all match, so we fold to all 0s.

File: llvm/test/CodeGen/X86/zext-sext.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/X86/zext-sext.ll:80]: There are no checks for the output

**Comment by fengfeng09** [comment]
> > > This optimization is not correct for 'undef'. I think it would be correct for 'poison', but SelectionDAG doesn't have 'poison' today.
> > > See also https://blog.regehr.org/archives/1837
> > 
> > 
> > I see. Thanks for your explanation, so zext undef to a zero is by designed, is this right?
> 
> For (zext undef), the zero extended bits of the result must be 0. The bits that started out as undef can be any value. We choose 0 for simplicity.
> 
> For (sext undef), the sign extended bits must have the same value as the original sign bit. Because of undef, the original sign bit can be either 0 or 1. Folding to undef wouldn't guarantee the extended bits all match, so we fold to all 0s.
Got it, thanks.

**Comment by fengfeng09** [comment]
> > > This optimization is not correct for 'undef'. I think it would be correct for 'poison', but SelectionDAG doesn't have 'poison' today.
> > > See also https://blog.regehr.org/archives/1837
> > 
> > 
> > I see. Thanks for your explanation, so zext undef to a zero is by designed, is this right?
> 
> For (zext undef), the zero extended bits of the result must be 0. The bits that started out as undef can be any value. We choose 0 for simplicity.
> 
> For (sext undef), the sign extended bits must have the same value as the original sign bit. Because of undef, the original sign bit can be either 0 or 1. Folding to undef wouldn't guarantee the extended bits all match, so we fold to all 0s.


Got it, thanks.

**Comment by fengfeng09** [comment]
This commit is not consistent with llvm undef refinement, so close it.

**Pull request closed without merging**